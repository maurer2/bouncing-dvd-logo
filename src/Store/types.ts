import type { colours } from './constants';
import type {
  START_GAME,
  TRIGGER_COLLISION,
  TOGGLE_PLAY_STATE,
  TOGGLE_SOUND,
  TRIGGER_COLLISION_END,
  SET_LAST_POSITION,
} from './actionTypes';

import type store from '.';

// store
export type Position = [x: number, y: number];
export type Colour = typeof colours[number];
export type Store = {
  lastPosition: Position | null;
  colours: {
    current: Colour;
    previous: Colour | null;
    available: Colour[];
  };
  isPaused: boolean;
  isPlayingSound: boolean;
  soundIsDisabled: boolean;
  collisionCount: number;
};
export type RootState = Store;
export type Dispatch = typeof store.dispatch;

// action types
export type StartGameAction = {
  type: typeof START_GAME;
};
export type TriggerCollisionAction = {
  type: typeof TRIGGER_COLLISION;
  payload: Colour;
};
export type TogglePlayStateAction = {
  type: typeof TOGGLE_PLAY_STATE;
};
export type ToggleSoundAction = {
  type: typeof TOGGLE_SOUND;
};
export type TriggerCollisionActionEnd = {
  type: typeof TRIGGER_COLLISION_END;
};
export type SetLastPositionAction = {
  type: typeof SET_LAST_POSITION;
  payload: Position;
};
export type Action =
  | StartGameAction
  | TriggerCollisionAction
  | TogglePlayStateAction
  | ToggleSoundAction
  | TriggerCollisionActionEnd
  | SetLastPositionAction;
