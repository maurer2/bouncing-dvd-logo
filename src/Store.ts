import { createContext } from 'react';

const colours = ['white', 'red', 'blue', 'yellow', 'fuchsia', 'lime'] as const;
type Colours = typeof colours[number];

export type StoreType = {
  colours: Colours[];
  soundIsDisabled: boolean;
  toggleSound: () => void;
};

const Store = createContext<StoreType | null>(null);

export default Store;
export {
  colours,
};
