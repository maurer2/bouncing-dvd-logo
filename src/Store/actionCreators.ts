import {
  TRIGGER_COLLISION_END,
  START_GAME,
  PAUSE_GAME,
  RESET_GAME,
  TRIGGER_COLLISION,
  TOGGLE_PLAY_STATE,
  TOGGLE_SOUND,
  SET_LAST_POSITION,
} from './actionTypes';
import type {
  Position,
  StartGameAction,
  PauseGameAction,
  ResetGameAction,
  TriggerCollisionAction,
  TogglePlayStateAction,
  ToggleSoundAction,
  TriggerCollisionActionEnd,
  SetLastPositionAction
} from './types';

import store from '.';

export function startGame(): StartGameAction {
  return {
    type: START_GAME,
  };
}

export function pauseGame(position: Position): PauseGameAction {
  return {
    type: PAUSE_GAME,
    payload: position,
  };
}

export function resetGame(): ResetGameAction {
  return {
    type: RESET_GAME,
  };
}

export function triggerCollision(): TriggerCollisionAction {
  const { current, available } = store.getState().colours;

  const coloursWithoutCurrent = available.filter((colour) => colour !== current);
  const randomColourIndex = Math.floor(Math.random() * coloursWithoutCurrent.length);
  const newColour = coloursWithoutCurrent[randomColourIndex];

  return {
    type: TRIGGER_COLLISION,
    payload: newColour,
  };
}

export function togglePlayState(): TogglePlayStateAction {
  return {
    type: TOGGLE_PLAY_STATE,
  };
}

export function toggleSound(): ToggleSoundAction {
  return {
    type: TOGGLE_SOUND,
  };
}

// temp
export function triggerCollisionEnd(): TriggerCollisionActionEnd {
  return {
    type: TRIGGER_COLLISION_END,
  };
}

export function setLastPosition(position: Position): SetLastPositionAction {
  return {
    type: SET_LAST_POSITION,
    payload: position,
  };
}
