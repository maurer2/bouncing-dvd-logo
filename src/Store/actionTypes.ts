export const START_GAME = 'START_GAME';
export const PAUSE_GAME = 'PAUSE_GAME';
export const RESET_GAME = 'RESET_GAME';

export const TRIGGER_COLLISION = 'TRIGGER_COLLISION';
export const TRIGGER_COLLISION_END = 'TRIGGER_COLLISION_END';
export const TOGGLE_PLAY_STATE = 'TOGGLE_PLAY_STATE';
export const TOGGLE_SOUND = 'TOGGLE_SOUND';
export const SET_LAST_POSITION = 'SET_LAST_POSITION';
