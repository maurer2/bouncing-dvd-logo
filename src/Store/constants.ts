export const colours = ['white', 'red', 'blue', 'yellow', 'fuchsia', 'lime'] as const;
