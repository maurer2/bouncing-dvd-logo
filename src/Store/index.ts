export {
  default,
} from './Store';

export {
  colours,
} from './Store';

export type {
  Colours, StoreType,
} from './types';
