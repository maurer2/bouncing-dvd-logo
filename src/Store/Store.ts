import type { Store } from './types';
import { colours } from './constants';

export const initialState: Store = {
  lastPosition: [0, 0],
  colours: {
    current: colours[0],
    previous: null,
    available: colours,
  },
  isPaused: false,
  soundIsDisabled: true,
  isPlayingSound: false,
  collisionCount: 0,
};
