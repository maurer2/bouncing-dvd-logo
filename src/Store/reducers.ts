import type { Store, Action, RootState } from './types';
import { initialState } from './Store';
import { colours } from './constants';

// eslint-disable-next-line @typescript-eslint/default-param-last
const reducers = (state: RootState = initialState, action: Action): Store => {
  switch (action.type) {
    case 'START_GAME': {
      return {
        ...state,
        isPaused: false,
        isPlayingSound: false,
        colours: {
          current: colours[0],
          previous: null,
          available: [...colours],
        },
        collisionCount: 0,
        lastPosition: null,
      };
    }
    case 'TOGGLE_PLAY_STATE': {
      return {
        ...state,
        isPaused: !state.isPaused,
      };
    }
    case 'TOGGLE_SOUND': {
      return {
        ...state,
        soundIsDisabled: !state.soundIsDisabled,
      };
    }
    case 'TRIGGER_COLLISION':
      return {
        ...state,
        isPlayingSound: !state.soundIsDisabled,
        collisionCount: state.collisionCount + 1,
        colours: {
          ...state.colours,
          current: action.payload,
          previous: state.colours.current,
        },
      };
    case 'TRIGGER_COLLISION_END':
      return {
        ...state,
        isPlayingSound: false,
      };
    case 'SET_LAST_POSITION':
      return {
        ...state,
        lastPosition: action.payload,
      };
    default: {
      return {
        ...state,
      };
    }
  }
};

export default reducers;
