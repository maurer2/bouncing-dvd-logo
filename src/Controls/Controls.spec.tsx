/* eslint-disable react/prop-types */
/* eslint-disable react/jsx-props-no-spreading */

import React from 'react';
import '@testing-library/jest-dom';
import { render } from '@testing-library/react';

import Store from '../Store';

import Component from './Controls';

describe('Components', () => {
  let storeValues = {
    colours: ['white', 'red', 'blue', 'yellow', 'fuchsia', 'lime'],
    soundIsDisabled: true,
    toggleSound: jest.fn(),
  };

  const StoreProvider = ({ children }) => (
    <Store.Provider value={storeValues}>
      {children}
    </Store.Provider>
  );

  const setup = (props) => render(
    <StoreProvider key={props?.key}>
      <Component {...props} />
    </StoreProvider>,
  );

  it('should render ', () => {
    const screen = setup({});

    expect(screen.getByTestId('controls')).toBeInTheDocument();
  });

  it('should match snapshot', () => {
    const screen = setup({});

    expect(screen.container.firstChild).toMatchSnapshot();
  });

  it('should have child elements', () => {
    const screen = setup({});

    expect(screen.getByTestId('controls')).toBeTruthy();

    expect(screen.getByTestId('controls-control')).toBeTruthy();
    expect(screen.getByRole('button')).toBeTruthy();

    expect(screen.getByTestId('controls-icon')).toBeTruthy();
  });

  it('should have sound off icon, when sound is disabled', () => {
    const screen = setup({});

    expect(screen.getByText('sound-off.svg')).toBeTruthy();
  });

  it('should have sound on icon, when sound is enabled', () => {
    storeValues = {
      ...storeValues,
      soundIsDisabled: false,
    };

    const screen = setup({ key: 'on' });

    expect(screen.getByText('sound-on.svg')).toBeTruthy();
  });
});
