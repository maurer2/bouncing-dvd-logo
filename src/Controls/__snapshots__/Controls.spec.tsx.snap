// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Components should match snapshot 1`] = `
<div
  class="Controlsstyles__Controls-kIyccD isEcOA"
  data-testid="controls"
>
  <button
    class="Controlsstyles__Control-fyrJBQ coIkqO"
    data-testid="controls-control"
    type="button"
  >
    <svg
      data-testid="controls-icon"
    >
      SVG markup
    </svg>
  </button>
</div>
`;
